kratos/utilities/pointer_map_communicator.h

```diff
@@ -0,0 +1,623 @@
+            // constructing an rank based GlobalPointersUnorderedMap will make unique keys list in GlobalPointersUnorderedMap
+            // which will result in lower communication for keys. These unique keys can be easily OMP parallelized
+            std::unordered_map<int, GlobalPointersUnorderedMap<TPointerDataType, std::vector<TValueDataType>>> non_local_map;
+
+            for (IndexType i = 0; i < rApplyProxy.mNonLocalGlobalPointerMapsVector.size(); ++i) {
+                const GlobalPointerMapType& current_gp_map = rApplyProxy.mNonLocalGlobalPointerMapsVector[i];
+                const DataVectorMapType& current_data_map = rApplyProxy.mNonLocalDataValueMapsVector[i];
+
+                for (const auto& r_gp_map_item : current_gp_map) {
+                    auto& current_rank_gp_map = non_local_map[r_gp_map_item.first];
+
+                    const auto& r_gp_vector = r_gp_map_item.second;
+                    const auto& r_data_vector = current_data_map.find(r_gp_map_item.first)->second;
+
+                    for (IndexType i = 0; i < r_gp_vector.size(); ++i) {
+                        current_rank_gp_map[r_gp_vector[i]].push_back(r_data_vector[i]);
+                    }
+                }
+            }
+
+            // compute the communication plan
+            const auto& colors = ComputeCommunicationPlan(non_local_map);
+
+            // perform send and receives to get and send remote data
+            for (const auto color : colors) {
+                if (color >= 0) {
+                    // In here
+                    //      1. We can do communication twice using default serializer for two map vectors.
+                    //              First map: (rank, gp)
+                    //              Second map: (rank, data_value)
+                    //      2. We can combine non_local_gp_map and non_local_data_map to one map, and do communication once
+                    //         after serializing the custom map.
+                    //              Current implementation
+                    //
+                    // Pros and Cons of 1:
+                    //      default serialization is used -> Pro
+                    //      OMP parallel loops cannot be used because received_gps_vector may not be unique -> Con
+                    //          This will be a performance hit if the lambda proxy is computationally expensive
+                    //      Communication need to be done twice once for gps, once for values -> Con
+                    // Pros and Const of 2:
+                    //      custom serialization will be used -> Con
+                    //      OMP parallel loops can be used in applying the proxy because the gps will be unique in the map -> Pro
+                    //      communication need to be done only once -> Pro
+                    //
+                    // I think the 2nd options outweighs the first one, so I implemented the second one. Suggestions are
```

Comment:
ok for the second option, although from my experience using MPI+OpenMP is not really effective (that is, if you have MPI in my experience it is best to use OpenMP=1 and more mpis)